packages/jsii-rosetta/lib/find-utils.ts

```diff
@@ -0,0 +1,86 @@
+  // package does not export that particular file.
+  const entryPoint = require.resolve(dependencyName, {
+    paths: [searchStart],
+  });
+
+  // Search up from the given directory, looking for a package.json that matches
+  // the dependency name (so we don't accidentally find stray 'package.jsons').
+  const depPkgJsonPath = await findPackageJsonUp(dependencyName, path.dirname(entryPoint));
+
+  if (!depPkgJsonPath) {
+    throw new Error(`Could not find dependency '${dependencyName}' from '${searchStart}'`);
+  }
+
+  return depPkgJsonPath;
+}
+
+/**
+ * Find the package.json for a given package upwards from the given directory
+ *
+ * (This code is duplicated among jsii/jsii-pacmak/jsii-reflect. Changes should be done in all
+ * 3 locations, and we should unify these at some point: https://github.com/aws/jsii/issues/3236)
+ */
+export async function findPackageJsonUp(packageName: string, directory: string) {
+  return findUp(directory, async (dir) => {
+    const pjFile = path.join(dir, 'package.json');
+    return (await fs.pathExists(pjFile)) && (await fs.readJson(pjFile)).name === packageName;
+  });
+}
+
+/**
+ * Find a directory up the tree from a starting directory matching a condition
+ *
+ * Will return `undefined` if no directory matches
+ *
+ * (This code is duplicated among jsii/jsii-pacmak/jsii-reflect. Changes should be done in all
+ * 3 locations, and we should unify these at some point: https://github.com/aws/jsii/issues/3236)
+ */
+export function findUp(directory: string, pred: (dir: string) => Promise<boolean>): Promise<string | undefined>;
+export function findUp(directory: string, pred: (dir: string) => boolean): string | undefined;
+// eslint-disable-next-line @typescript-eslint/promise-function-async
+export function findUp(
+  directory: string,
+  pred: ((dir: string) => boolean) | ((dir: string) => Promise<boolean>),
+): Promise<string | undefined> | string | undefined {
+  const result = pred(directory);
+  if (isPromise(result)) {
+    return result.then((thisDirectory) => (thisDirectory ? directory : recurse()));
+  }
+
+  return result ? directory : recurse();
+
+  function recurse() {
+    const parent = path.dirname(directory);
+    if (parent === directory) {
+      return undefined;
+    }
+    return findUp(parent, pred as any);
+  }
+}
+
+function isPromise<A>(x: A | Promise<A>): x is Promise<A> {
+  return typeof x === 'object' && (x as any).then;
+}
+
+/**
+ * Whether the given dependency is a built-in
+ *
+ * Some dependencies that occur in `package.json` are also built-ins in modern Node
+ * versions (most egregious example: 'punycode'). Detect those and filter them out.
+ */
+export function isBuiltinModule(depName: string) {
+  // eslint-disable-next-line @typescript-eslint/no-require-imports,@typescript-eslint/no-var-requires
+  const { builtinModules } = require('module');
+  return (builtinModules ?? []).includes(depName);
+}
```

Comment:
AFAIK, the behavior of `node` module resolution is that, unless the module specifier (in `require(moduleSpecifier)` or `import from moduleSpecifier`) starts with `node:` (at which point it will ALWAYS resolve to a built-in module), it'll start by trying to resolve from external library search paths, and ONLY if not found will it resort to built-ins.

This makes me feel like this behavior is, in fact, incorrect in this respects.

Reply:
It doesn't, though.

```
$ npm init -y
$ npm install punycode

added 1 package, and audited 2 packages in 601ms

found 0 vulnerabilities

$ node -p "require.resolve('punycode')"
punycode                <--- indicating a built-in module
```

Reply:
Ah you're correct, the module resolution STARTS with "if this is a built-in, return the built-in". My bad, sorry.